Add checked/dynamic casts for OSObjectPtr

ddkilzer opened the pull request:
#### d9672f0e75352880994a81f213bfd3bd560141b0
<pre>
Add checked/dynamic casts for OSObjectPtr
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=297639">https://bugs.webkit.org/show_bug.cgi?id=297639</a>&gt;
&lt;<a href="https://rdar.apple.com/158736926">rdar://158736926</a>&gt;

Reviewed by Darin Adler.

OSObject types behave like CF objects in C/C++, but they behave like
Cocoa objects in Objective-C/C++, including opting into ARC.  Unlike
toll-free bridged types, OSObject CF types and the Cocoa types aren&apos;t
defined at the same time, which complicates implementing a shared
solution for both.

For example, an OSObject of type `foo_bar` will be defined like this in
plain C:

  struct foo_bar;
  typedef struct foo_bar *foo_bar_t;

Note that some implementations (like libdispatch) have separate
definitions for C++ with an &quot;_s&quot; suffix:

  typedef struct dispatch_object_s { ... } *dispatch_object_t;

On the other hand, an OSObject of type `foo_bar` will be defined like
this in Objective-C/C++:

  @protocol OS_foo_bar &lt;NSObject&gt; @end
  typedef NSObject&lt;OS_foo_bar&gt; * __attribute__((objc_independent_class)) foo_bar_t;

OSObject &quot;class&quot; inheritance is implemented in Objective-C/C++ by making
the @protocol inherit from the @protocol of the superclass, or inherit
from NSObject if it&apos;s a top-level class.  There is no concept of
inheritence in plain C, although libdispatch does provide inheritance of
struct types in C++:

  typedef struct dispatch_group_s : public dispatch_object_s {} *dispatch_group_t;

However, most OSObject types don&apos;t implement separate C++ definitions,
so that can&apos;t be relied upon to check inheritance.  (These C++
definitions are also not available in Objective-C++ sources, so it would
require implementation of two different type checks, which is less than
ideal.)

Instead, to check whether a given object is of a specific OSObject type,
the Objective-C +[Class conformsToProtocol:] method (or
class_conformsToProtocol() in C) may be used.  Conveniently, any
OSObject that&apos;s currently a plain C/C++ type may be treated as a
CFTypeRef, bridge-cast to `id`, and then used to call either
+conformsToProtocol: or class_conformsToProtocol().

The challenge with implemeting isOSObject&lt;T&gt;() is that an OSObject type,
such as foo_bar_t, is defined differently in plain C++ and
Objective-C++, as noted above.

To make it possible to use a single implementation of isOSObject&lt;T&gt;()
for a type (regardless of the language), OSObjectTypeCastTraits structs
are defined so that `T` is mapped to a common type that&apos;s shared between
C++ and Objective-C++.

If `T` is an OSObject of type `foo_bar_t`, the trait is defined like
this in plain C++ (with the typedef resolved to the underlying type):

  template&lt;&gt; struct OSObjectTypeCastTraits&lt;struct foo_bar*&gt; { using BaseType = struct foo_bar*; };

But it is defined like this in Objective-C++:

  template&lt;&gt; struct OSObjectTypeCastTraits&lt;NSObject&lt;OS_foo_bar&gt; *&gt; { using BaseType = struct foo_bar*; };

While redundant for plain C++, this allows us to map `foo_bar_t` to the
same type no matter which language is used at compile time, and thus
only implement isOSObject&lt;&gt;() once per OSObject type:

  template&lt;&gt; bool isOSObject&lt;OSObjectTypeCastTraits&lt;foo_bar_t&gt;::BaseType&gt;(CFTypeRef object) {
      Class cls = object_getClass(bridge_id_cast(object));
      return class_conformsToProtocol(cls, objc_getProtocol(ProtocolString));
  }

The only downside to this approach is that we must embed some knowledge
of the plain C/C++ struct type when constructing OSObjectTypeCastTraits
for Objective-C++, although it&apos;s only a pre-declared struct.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
- Add TypeCastsOSObject.h to project.
* Source/WTF/wtf/cf/TypeCastsCF.h:
- Add `using WTF::bridgeCFCast` statement.
- Move `using WTF::bridge_id_cast` statement here from TypeCastsCocoa.h.
(WTF::bridgeCFCast): Add.
- Add convenience function to cast from CFTypeRef to id without
  reference churn.  Works in both plain C++ and Objective-C++.
(WTF::bridge_id_cast):
- Move from TypeCastsCocoa.h and make work for CF in plain C++.
* Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
- Include &lt;wtf/cf/TypeCastsCF.h&gt; for bridge_id_cast().
- Move `using WTF::bridge_id_cast` statement to TypeCastsCF.h.
* Source/WTF/wtf/darwin/TypeCastsOSObject.h: Add.
(WTF::isOSObject): Add.
(WTF::osObjectCast): Add.
(WTF::dynamicOSObjectCast): Add.
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::DNSResolveQueueCFNet::performDNSLookup):
- Add dynamic cast.
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(extractResolutionReport):
- Add dynamic casts.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
- Add TypeCastsOSObjectCF.cpp, TypeCastsOSObjectCocoa.mm and
  TypeCastsOSObjectCocoaARC.mm to the project.
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCF.cpp: Add.
(TestWebKitAPI::TEST(TypeCastsOSObjectCF, osObjectCast)): Add.
(TestWebKitAPI::TEST(TypeCastsOSObjectCF, dynamicOSObjectCast)): Add.
(TestWebKitAPI::TEST(TypeCastsOSObjectCF, dynamicOSObjectCast_OSObjectPtr)): Add.
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCocoa.mm: Add.
(TestWebKitAPI::TEST(TYPE_CASTS_OSOBJECT_PTR_TEST_NAME, osObjectCast_from_id)): Add.
(TestWebKitAPI::TEST(TYPE_CASTS_OSOBJECT_PTR_TEST_NAME, dynamicOSObjectCast_from_OSObjectPtr)): Add.
(TestWebKitAPI::TEST(TYPE_CASTS_OSOBJECT_PTR_TEST_NAME, dynamicOSObjectCast_from_id)): Add.
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCocoaARC.mm: Add.

Canonical link: <a href="https://commits.webkit.org/304847@main">https://commits.webkit.org/304847@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/09d0aacbd6c12d5c2a626e2d9f7eb29cb97b54be

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136639 "17 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8998 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47926 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144363 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89612 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/484df2b2-dcb2-429d-bc85-5de835644152) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138511 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9699 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8844 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104482 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/75059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8f337c98-b7ec-401c-967b-36fff37e03a5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139584 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7082 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122430 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85319 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8195cc7d-5c69-4ddd-a5de-52d802dca8c4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6726 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4409 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4959 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128600 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116041 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40619 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147122 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135125 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8682 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41193 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112836 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8700 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7303 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113166 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6649 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118723 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62789 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21070 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8730 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36777 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167904 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8450 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72296 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43806 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8670 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8522 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/darwin/TypeCastsOSObject.h:
```diff
@@ -0,0 +1,224 @@
+    if (!object)
+        return nullptr;
+
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(is_osObject<typename OSObjectTypeCastTraits<T>::BaseType>(object));
```
I don't think this has security implication? We're asserting if the type check failed.

ddkilzer replied:
Will fix.  Copied from `dynamic_objc_cast()`.

rniwa commented on Source/WTF/wtf/darwin/TypeCastsOSObject.h:
```diff
@@ -0,0 +1,224 @@
+
+#ifdef __OBJC__
+
+template<typename T> inline bool is_osObject(id object)
```
I think we want to use camelCase. There is a bug / PR to update retain ptr functions: https://github.com/WebKit/WebKit/pull/45093

ddkilzer replied:
Okay, I'll change it to `isOSObject`.

darinadler commented on Source/WTF/wtf/darwin/TypeCastsOSObject.h:
```diff
@@ -0,0 +1,224 @@
+#define WTF_OS_OBJECT_NETWORK_TYPES(M) \
+    M(nw_endpoint) \
+    M(nw_resolution_report) \
+    M(nw_path)
```
```suggestion
    M(nw_path) \
    M(nw_resolution_report)
```

darinadler commented on Source/WTF/wtf/darwin/TypeCastsOSObject.h:
```diff
@@ -0,0 +1,224 @@
+    WTF_OS_OBJECT_DISPATCH_TYPES(M) \
+    WTF_OS_OBJECT_NETWORK_TYPES(M)
+
+// Because ARC enablement is a compile-time choice, and we compile this header
+// both ways, we need a separate copy of our code when ARC is enabled.
+#if __has_feature(objc_arc)
+#define dynamicOSObjectCast dynamicOSObjectCastARC
+#define osObjectCast osObjectCastARC
+#endif
```
I'd be tempted to reverse this since ARC is the future:
```suggestion
// Because ARC enablement is a compile-time choice, and we compile this header
// both ways, we need a separate copy of our code when ARC is enabled.
#if !__has_feature(objc_arc)
#define dynamicOSObjectCast dynamicOSObjectCastNonARC
#define osObjectCast osObjectCastNonARC
#endif
```

ddkilzer replied:
IMO, these macros are a hack either way they're defined, and the "NonARC" variants won't realistically go away because we'll always have CF types (always non-ARC) or (in rare cases) use NS types in non-Objective-C[++] sources.

I think a better way to do this is to use `struct ARCEnabled` added by @cdumez in `Source/WTF/wtf/TypeTraits.h` to differentiate `RetainPtr<>` and `OSObjectPtr<>` at build-time by adding a second template parameter.

darinadler reviewed (approved):
Wish there was a way to make this more generic and deduce what to do from underlying types so we could share more of `RetainPtr` and overload the different casting function templates instead of having to use distinct names. I hope at least it fails to compile if you use the wrong function rather than compiling and doing the wrong thing at runtime.

iidmsa commented on Source/WTF/wtf/darwin/TypeCastsOSObject.h:
```diff
@@ -0,0 +1,224 @@
+// in Objective-C++ vs. C++, so these Objective-C method definitions do not create an ODR violation
+// with the C++ method definitions below.
+
+template<typename T, typename U, typename = std::enable_if_t<!std::is_same_v<U, T>>> inline OSObjectPtr<T> dynamicOSObjectCast(OSObjectPtr<U>&& object)
```
Would this work? 
```suggestion
template<typename T, typename U>
inline OSObjectPtr<T> dynamicOSObjectCast(OSObjectPtr<U>&& object)
    requires (!std::same_as<U, T>)
```

ddkilzer replied:
Yes, thanks!

Commit:
Add checked/dynamic casts for OSObjectPtr
<https://bugs.webkit.org/show_bug.cgi?id=297639>
<rdar://158736926>

Reviewed by Darin Adler.

OSObject types behave like CF objects in C/C++, but they behave like
Cocoa objects in Objective-C/C++, including opting into ARC.  Unlike
toll-free bridged types, OSObject CF types and the Cocoa types aren't
defined at the same time, which complicates implementing a shared
solution for both.

For example, an OSObject of type `foo_bar` will be defined like this in
plain C:

  struct foo_bar;
  typedef struct foo_bar *foo_bar_t;

Note that some implementations (like libdispatch) have separate
definitions for C++ with an "_s" suffix:

  typedef struct dispatch_object_s { ... } *dispatch_object_t;

On the other hand, an OSObject of type `foo_bar` will be defined like
this in Objective-C/C++:

  @protocol OS_foo_bar <NSObject> @end
  typedef NSObject<OS_foo_bar> * __attribute__((objc_independent_class)) foo_bar_t;

OSObject "class" inheritance is implemented in Objective-C/C++ by making
the @protocol inherit from the @protocol of the superclass, or inherit
from NSObject if it's a top-level class.  There is no concept of
inheritence in plain C, although libdispatch does provide inheritance of
struct types in C++:

  typedef struct dispatch_group_s : public dispatch_object_s {} *dispatch_group_t;

However, most OSObject types don't implement separate C++ definitions,
so that can't be relied upon to check inheritance.  (These C++
definitions are also not available in Objective-C++ sources, so it would
require implementation of two different type checks, which is less than
ideal.)

Instead, to check whether a given object is of a specific OSObject type,
the Objective-C +[Class conformsToProtocol:] method (or
class_conformsToProtocol() in C) may be used.  Conveniently, any
OSObject that's currently a plain C/C++ type may be treated as a
CFTypeRef, bridge-cast to `id`, and then used to call either
+conformsToProtocol: or class_conformsToProtocol().

The challenge with implemeting isOSObject<T>() is that an OSObject type,
such as foo_bar_t, is defined differently in plain C++ and
Objective-C++, as noted above.

To make it possible to use a single implementation of isOSObject<T>()
for a type (regardless of the language), OSObjectTypeCastTraits structs
are defined so that `T` is mapped to a common type that's shared between
C++ and Objective-C++.

If `T` is an OSObject of type `foo_bar_t`, the trait is defined like
this in plain C++ (with the typedef resolved to the underlying type):

  template<> struct OSObjectTypeCastTraits<struct foo_bar*> { using BaseType = struct foo_bar*; };

But it is defined like this in Objective-C++:

  template<> struct OSObjectTypeCastTraits<NSObject<OS_foo_bar> *> { using BaseType = struct foo_bar*; };

While redundant for plain C++, this allows us to map `foo_bar_t` to the
same type no matter which language is used at compile time, and thus
only implement isOSObject<>() once per OSObject type:

  template<> bool isOSObject<OSObjectTypeCastTraits<foo_bar_t>::BaseType>(CFTypeRef object) {
      Class cls = object_getClass(bridge_id_cast(object));
      return class_conformsToProtocol(cls, objc_getProtocol(ProtocolString));
  }

The only downside to this approach is that we must embed some knowledge
of the plain C/C++ struct type when constructing OSObjectTypeCastTraits
for Objective-C++, although it's only a pre-declared struct.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
- Add TypeCastsOSObject.h to project.
* Source/WTF/wtf/cf/TypeCastsCF.h:
- Add `using WTF::bridgeCFCast` statement.
- Move `using WTF::bridge_id_cast` statement here from TypeCastsCocoa.h.
(WTF::bridgeCFCast): Add.
- Add convenience function to cast from CFTypeRef to id without
  reference churn.  Works in both plain C++ and Objective-C++.
(WTF::bridge_id_cast):
- Move from TypeCastsCocoa.h and make work for CF in plain C++.
* Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
- Include <wtf/cf/TypeCastsCF.h> for bridge_id_cast().
- Move `using WTF::bridge_id_cast` statement to TypeCastsCF.h.
* Source/WTF/wtf/darwin/TypeCastsOSObject.h: Add.
(WTF::isOSObject): Add.
(WTF::osObjectCast): Add.
(WTF::dynamicOSObjectCast): Add.
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::DNSResolveQueueCFNet::performDNSLookup):
- Add dynamic cast.
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(extractResolutionReport):
- Add dynamic casts.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
- Add TypeCastsOSObjectCF.cpp, TypeCastsOSObjectCocoa.mm and
  TypeCastsOSObjectCocoaARC.mm to the project.
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCF.cpp: Add.
(TestWebKitAPI::TEST(TypeCastsOSObjectCF, osObjectCast)): Add.
(TestWebKitAPI::TEST(TypeCastsOSObjectCF, dynamicOSObjectCast)): Add.
(TestWebKitAPI::TEST(TypeCastsOSObjectCF, dynamicOSObjectCast_OSObjectPtr)): Add.
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCocoa.mm: Add.
(TestWebKitAPI::TEST(TYPE_CASTS_OSOBJECT_PTR_TEST_NAME, osObjectCast_from_id)): Add.
(TestWebKitAPI::TEST(TYPE_CASTS_OSOBJECT_PTR_TEST_NAME, dynamicOSObjectCast_from_OSObjectPtr)): Add.
(TestWebKitAPI::TEST(TYPE_CASTS_OSOBJECT_PTR_TEST_NAME, dynamicOSObjectCast_from_id)): Add.
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCocoaARC.mm: Add.

Canonical link: https://commits.webkit.org/304847@main
